## Feature Scaling
The scales of variables can have a very large impact on the results of the model. For instance, consider this example of employee salaries:

| ID    | Years of Experience | Salary  |
| :---: | :-----------------: | :-----: |
| 01    |     5               | $70,000 |
| 02    |     6               | $60,000 |
| 03    |     1               | $52,000 |

In [1]:
import pandas as pd

salaries = {
    'ID':['01','02','03'],
    'Salary':[70000,60000,52000],
    'Years of Experience':[5,6,1]
}

df = pd.DataFrame(salaries, index=salaries['ID']).drop('ID', axis=1)
df

,Salary,Years of Experience
01,70000,5
02,60000,6
03,52000,1


Let's compare the salary and the years of experience for each employee to the average. First, find the averages:
* Average years of experience: 3.333 years
* Average salary: \$60,667

We can compare employee 1 to the average using Pythagorean Theorem:

$$d = \sqrt{(70,000 - 60,667)^2 + (5 - 3.333)^2}$$

Which term is going to have the biggest effect on the distance? It would be the salary. The resulting distance is,

$$d = \sqrt{(70,000 - 60,667)^2 + (5 - 3.333)^2} = \sqrt{87,104,889 + 2.7789} = 9,333$$

Note that this distance is exactly the same as the distance between salaries. The distance between years of experience is small, so dwarfed in comparison, contributing very little to the model. Perhaps that is fine and as it should be. But now look at employee 2:

$$d = \sqrt{(60,000 - 60,667)^2 + (6 - 3.333)^2}$$

In this example, the salary is very close to the mean and the years of experience is further off. In this case, the years of experience is more significant. But the calculation still overemphasizing the salary:

$$d = \sqrt{(60,000 - 60,667)^2 + (6 - 3.333)^2} = \sqrt{444,889 + 7.1129} = 667$$

Note again that this distance is exactly the same as the distance between salaries. In this case, the difference in salaries should be less emphasized since they are nearly the same and the years of experience should have been emphasized more, but it wasn't. This is because of the square in the distance formula - large values are increased exponentially and small numbers are dwarfed just as fast.

The scale is throwing us off, so we look at __feature scaling__. There are two primary methods of feature scaling:
1. Standardization
$$\hat{x} = \frac{x-\bar{x}}{s}$$
2. Normalization
$$\hat{x} = \frac{x-x_{min}}{x_{max}-x_{min}}$$

Standardization will generally give a number in the range [-3,3], while normalization will always give a result between [0,1].

Let's see how each method affects the data.

In [2]:
# Standardized
def standardize_df(x):
    return (x-x.mean())/(x.std(ddof=1))

standardize_df(df)

,Salary,Years of Experience
01,1.034910,0.377964
02,-0.073922,0.755929
03,-0.960988,-1.133893


In [3]:
def normalize_df(x):
    return (x-x.min())/(x.max()-x.min())

normalize_df(df)

,Salary,Years of Experience
01,1.000000,0.8
02,0.444444,1.0
03,0.000000,0.0


With the data on similar scales, let's look at the distances for employee 2 once again.
* Average Normalized Salary: 0.48148
* Average Normalized Years of Experience: 0.6

$$d_1 = \sqrt{(1.0 - 0.48148)^2 + (0.8 - 0.6)^2} = 0.5558$$

$$d_2 = \sqrt{(0.44444 - 0.48148)^2 + (1.0 - 0.6)^2} = 0.4017$$

This time, the distance more acurately reflects (and balances) the differences in each component.

We looked at normalization and standardization. There are other methods for scaling features. However, these are the two most common, so the ones we will focus on in the next few semesters.